File: Data Structures & Algorithms/add-two-numbers/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        sumList = ListNode(0)
        head = sumList
        while(l1 and l2):
            # sum = l1.val + l2.val + carry
            # if(sum >= 10):
            #     carry = 1
            #     sumList.next = ListNode(sum - 10)
            # else:
            #     sumList.next = ListNode(sum)
            #     carry = 0
            sum, carry = self.sumNodes(l1.val, l2.val, carry)
            sumList.next = ListNode(sum)
            l1 = l1.next
            l2 = l2.next
            sumList = sumList.next
        while(l1):
            sum, carry = self.sumNodes(l1.val, 0, carry)
            sumList.next = ListNode(sum)
            sumList = sumList.next
            l1 = l1.next
        while(l2):
            sum, carry = self.sumNodes(0, l2.val, carry)
            sumList.next = ListNode(sum)
            sumList = sumList.next
            l2 = l2.next
        if(carry and l1 is None and l2 is None):
            sumList.next = ListNode(carry)
        return head.next
    
    def sumNodes(self, num1, num2, carry):
        sum = num1 + num2 + carry
        if(sum >= 10):
            carry = 1
            sum = sum - 10
        else:
            carry = 0
        
        return [sum, carry]

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {
        if(!list1 && !list2){
            return list1;
        }
        else if(!list1 && list2){
            return list2
        }
        else if(list1 && !list2){
            return list1
        }
        // let firstVal = list1.val < list2.val ? list1 : list2;
        let curr = new ListNode(null);
        let head = curr;
        // 
        while(list1 && list2) {
            if(list1.val <= list2.val){
                curr.next = new ListNode(list1.val);
                list1 = list1.next;
            } else {
                curr.next = new ListNode(list2.val);
                list2 = list2.next;
            }
            curr = curr.next
        }
        if(list1) curr.next = list1
        if(list2) curr.next = list2
        return head.next;
    }
}
